src/pages/home/report/withReportAndReportActionOrNotFound.js

```diff
@@ -0,0 +1,147 @@
+import * as ReportUtils from '../../../libs/ReportUtils';
+import * as ReportActionsUtils from '../../../libs/ReportActionsUtils';
+import * as Report from '../../../libs/actions/Report';
+import compose from '../../../libs/compose';
+import withWindowDimensions from '../../../components/withWindowDimensions';
+
+export default function (WrappedComponent) {
+    const propTypes = {
+        /** The HOC takes an optional ref as a prop and passes it as a ref to the wrapped component.
+         * That way, if a ref is passed to a component wrapped in the HOC, the ref is a reference to the wrapped component, not the HOC. */
+        forwardedRef: PropTypes.func,
+
+        /** The report currently being looked at */
+        report: reportPropTypes,
+
+        /** Array of report actions for this report */
+        reportActions: PropTypes.shape(reportActionPropTypes),
+
+        /** The policies which the user has access to */
+        policies: PropTypes.objectOf(
+            PropTypes.shape({
+                /** The policy name */
+                name: PropTypes.string,
+
+                /** The type of the policy */
+                type: PropTypes.string,
+            }),
+        ),
+
+        /** Route params */
+        route: PropTypes.shape({
+            params: PropTypes.shape({
+                /** Report ID passed via route */
+                reportID: PropTypes.string,
+
+                /** ReportActionID passed via route */
+                reportActionID: PropTypes.string,
+            }),
+        }).isRequired,
+
+        /** Beta features list */
+        betas: PropTypes.arrayOf(PropTypes.string),
+
+        /** Indicated whether the report data is loading */
+        isLoadingReportData: PropTypes.bool,
+
+        /** Is the window width narrow, like on a mobile device? */
+        isSmallScreenWidth: PropTypes.bool.isRequired,
+    };
+
+    const defaultProps = {
+        forwardedRef: () => {},
+        reportActions: {},
+        report: {
+            isLoadingReportActions: true,
+        },
+        policies: {},
+        betas: [],
+        isLoadingReportData: true,
+    };
+
+    // eslint-disable-next-line rulesdir/no-negated-variables
+    function WithReportAndReportActionOrNotFound(props) {
+        // For small screen, we don't call openReport API when we go to a sub report page by deeplinnk
+        // So we need to call openReport here for small screen
+        const firstRef = useRef(true);
+        useEffect(() => {
+            if (!firstRef.current) {
+                return;
+            }
+            firstRef.current = false;
+            if (props.isSmallScreenWidth) {
+                Report.openReport(props.route.params.reportID);
+            }
+        }, [props.isSmallScreenWidth, props.route.params.reportID]);
+
+        const isLoadingReport = props.isLoadingReportData && (_.isEmpty(props.report) || !props.report.reportID);
```

Comment:
I understand these checks are complex, but we generally try to early return. Is it possible to structure the code:

```
// Perform all the loading checks
 const isLoadingReport  = ...;
 const isLoadingReportAction = ....;

 if ((isLoadingReport || isLoadingReportAction) && !shouldHideReport) {
    return <FullscreenLoadingIndicator />;
}


// Perform the access/not found checks
const shouldHideReport
if (shouldHideReport || _.isEmpty(reportAction)) {
    return <NotFoundPage />;
}



```